src/basic/time-util.c

```diff
@@ -1493,3 +1493,34 @@ int time_change_fd(void) {
 
         return TAKE_FD(fd);
 }
+
+/*
+ * Clamp times *a and *b so that they are in range [min, max]
+ * while still keeping the relative proportion between them.
+ * Returns true if some clamping has been performed, in which case
+ * *a and *b will be updated with the new values.
+ */
+bool clamp_time_interval(usec_t *a, usec_t *b, usec_t min, usec_t max) {
```

Comment:
i am not sure i grok why we need this function... the ratio is stored as ratio anyway, hence when the period is altered the ratio doesn't have to be changed.

We should simply clamp the period early (i.e. while parsing) and then at the time we write to cgroupfs calculate the time value from the period and the ratio. 

This also means the ratio never has to be clamped. 

i.e. if you say 76% quota, then it means mostly the same regardless if you operate with a 100ms period or a 500ms period: as long as we calculate 76% of the period at the instant we write to cgroupfs we are fine.

Also note quota above 100% is completely OK, as the value is counted per CPU, hence 200% means 2 cpus full time.

Reply:
So we need this because both `period` and `quota * period` (not sure what to call this one... bandwidth?) need to be within the `[1ms, 1s]` interval.

So if I have `CPUQuota=40%` and `CPUQuotaPeriodSec=10ms`, then the first field will be set to 4ms and everything is fine (no clamping needed.) However, if I have the same `CPUQuota=40%` but `CPUQuotaPeriodSec=2ms`, then this needs clamping, even though the period is in the allowed interval! But `quota * period = 0.8ms` is out of it, so this needs to be clamped to `1ms` (for the `quota * period` product, minimum in the range) and `2.5ms` for the period, in order to keep the ratio of 40%.

Does this make sense now?

Reply:
hmm, so this is just about ensuring the value never becomes 0 after scaling?

I think to deal with that a scaling that rounds up would make sense (which I think would be wise anyway). I.e.

`(quota * period + 999999) / 1000000`

With that logic we'd never go to zero, given that neither quota nor period can be zero...

Reply:
No, it's not about zero... It's about making sure that the calculated slice `quota * period` is _also_ in the `[1ms, 1s]` range. Otherwise, writing to the cgroup file (at least in cgroupv2) fails.

Like, with this example, `CPUQuota=40%` and `CPUQuotaPeriodSec=2ms`, the calculated slice `quota * period = 0.8ms`, but setting `cpu.max` to `800 2000` fails. So this needs to be scaled up to `1000 2500`, which is the smallest possible period that satisfies both `period` _and_ `quota * period` being within the valid interval.

I hope this makes sense?

I'll definitely add more comments here...

I think other than the check for division by zero, this function should be OK. But I'm open to using a slightly different prototype, perhaps sending the quota as is (a percentage) and then just returning the calculated period back... But I think this prototype works fine as is, it's pretty generic and avoids us having to calculate that more than once...

What do you think would be a good name for `quota * period`? I thought of `bandwidth` or `slice`...

One part that is confusing in this code is this other function (`cgroup_cpu_adjust_quota_period`) which takes `quota` as an argument (by reference) but returns `quota * period` (i.e. `slice` or `bandwidth`) in that same argument... Perhaps that could use some improvement.

Cheers,
Filipe

Reply:
Ping? Please see this comment here, I think that's the main point of contention that needs clarification now. Thanks!

Reply:
so the range is actually only 1ms…1s for the period but for quota*period there is no upper limit, only 1ms lower limit.

now, i am not entirely sure what the right approach to clamping here is if these limits are breached. if i understand you correctly then you opted for a scheme where the quota is kept in effect and the period and the period adjusted if needed. i.e. you consider quota more relevant than period. I figure that makes sense indeed.

Reply:
> What do you think would be a good name for quota * period? I thought of bandwidth or slice...

the kernel calls that quota too...

maybe call it "quota_absolute" or so?

Reply:
Ah indeed I didn't realize that setting quota_absolute > 1s actually worked... That makes this a bit more complicated, but I'll take a look and see if I can make it nicer.

(And yes, I like quota_absolute, will use that.)